crypto/fips/self_test.c

```diff
@@ -0,0 +1,349 @@
+ */
+int SELF_TEST_keygen_pairwise_test_rsa(RSA *rsa)
+{
+    int ret = 1;
+    unsigned int ciphertxt_len;
+    unsigned char *ciphertxt;
+    const unsigned char plaintxt[16] = {0};
+    unsigned char decoded[256];
+    unsigned int decoded_len;
+    unsigned int plaintxt_len = (unsigned int)sizeof(plaintxt_len);
+    int padding = RSA_PKCS1_PADDING;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT,
+                            SELF_TEST_DESC_PCT_RSA_PKCS1);
+
+    ciphertxt_len = RSA_size(rsa);
+    ciphertxt = OPENSSL_zalloc(ciphertxt_len);
+    if (ciphertxt == NULL) {
+        goto err;
+    }
+
+    ciphertxt_len = RSA_public_encrypt(plaintxt_len, plaintxt, ciphertxt, rsa,
+                                       padding);
+    if (ciphertxt_len <= 0)
+        goto err;
+    if (ciphertxt_len == plaintxt_len
+            && memcmp(decoded, plaintxt, plaintxt_len) == 0) {
+        goto err;
+    }
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, ciphertxt);
+
+    decoded_len = RSA_private_decrypt(ciphertxt_len, ciphertxt, decoded, rsa,
+                                      padding);
+    if (decoded_len != plaintxt_len
+            || memcmp(decoded, plaintxt,  decoded_len) != 0) {
+        goto err;
+    }
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    OPENSSL_free(ciphertxt);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_ecdsa(EC_KEY *eckey)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    ECDSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_ECDSA);
+
+    sig = ECDSA_do_sign(dgst, dgst_len, eckey);
+    if (sig == NULL)
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, dgst);
+
+    if (ECDSA_do_verify(dgst, dgst_len, sig, eckey) != 1)
+        goto err;
+
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    ECDSA_SIG_free(sig);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_dsa(DSA *dsa)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    DSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_DSA);
+
+    sig = DSA_do_sign(dgst, (int)dgst_len, dsa);
+    if (sig == NULL)
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, dgst);
+
+    if (DSA_do_verify(dgst, dgst_len, sig, dsa) != 1)
+        goto err;
+
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    DSA_SIG_free(sig);
+    return ret;
+}
+
+#define INTEGRITY_BUF_SIZE (4 * 1024)
+
+/*
+ * Calculate the HMAC SHA256 of data read using a BIO and read_cb, and verify
+ * the result matches the expected value.
+ * Return 1 if verified, or 0 if it fails.
+ */
+static int verify_integrity(BIO *bio, BIO_READ_CB read_cb,
```

Comment:
Are we actually planning to pull BIO code into the FIPS module?

Reply:
The module needs to be able to read itself (i.e. the shared object "file" it was loaded from).  This can be done by passing BIO calls in, using stdio, ....  there are many possibilities.

Which is best?  I'm not sure.

At the Edinburgh meeting, no file system devices were mentioned (which kind of rules out stdio).  BIO seemed natural since it can handle files, memory mapped, blocks in flash, serial roms, ... pretty much whatever.

No BIO code needs to be in the FIPS module -- just the appropriate callbacks and an opaque pointer.  If any BIO does end up pulled in, something has gone terribly wrong.

Reply:
Yep they are all callbacks..